File: algorithm/algs4/Stack.rb
# $ java-algs4 edu.princeton.cs.algs4.Stack < tobe.txt
# to be not that or be (2 left on stack)
# $ ruby Stack.rb < tobe.txt
# to be not that or be (2 left on stack)

module LinkedList
  class Node
    attr_accessor :item, :next_node

    def initialize(item, next_node)
      @item = item
      @next_node = next_node
    end
  end

  class Stack
    include Enumerable

    def initialize
      @first = nil
      @size = 0
    end

    def push(item)
      # Ruby is evaluated from right to left
      @first = Node.new(item, @first)
      @size += 1
    end
    alias_method :<<, :push
    
    def pop
      raise "Stack is empty." if empty?
      item = @first.item
      @first = @first.next_node
      @size -= 1
      item
    end
    
    def empty?
      @first.nil?
    end
    
    def size
      @size
    end
 
    def peek
      raise "Stack is empty" if empty?
      @first.item
    end

    def each
      current = @first
      while current.next_node != nil
        yield current.item
        current = current.next_node
      end
      yield current.item
    end

    def to_s
      elements = []
      self.each do |i|
        elements << i
      end
      elements.join(' ')
    end
  end
end

stack = LinkedList::Stack.new

ARGF.each_line(' ') do |line|
  item = line.strip
  if item != '-'
    stack << item
  elsif !stack.empty?
    print stack.pop + ' '
  end
end

print "(#{stack.size} left on stack)\n"

File: algorithm/algs4/Stack.go
// $ java-algs4 edu.princeton.cs.algs4.Stack < tobe.txt
// to be not that or be (2 left on stack)
// $ go run Stack.go < tobe.txt
// to be not that or be (2 left on stack)

package main

import (
	"bufio"
	"errors"
	"fmt"
	"os"
)

type Node struct {
	item interface{}
	next *Node
}

type Stack struct {
	first *Node
	size  int
}

func (s *Stack) Push(item interface{}) {
	s.first = &Node{item, s.first}
	s.size++
}

func (s *Stack) Pop() (item interface{}) {
	if s.size > 0 {
		item, s.first = s.first.item, s.first.next
		s.size--
		return item
	}
	return nil
}

func (s *Stack) IsEmpty() bool {
	return s.first == nil
}

func (s *Stack) Len() int {
	return s.size
}

func (s *Stack) Peek() (item interface{}, err error) {
	if s.IsEmpty() {
		err = errors.New("Stack is empty.")
		return nil, err
	}
	return s.first.item, nil
}

func (s *Stack) ToString() string {
	var a string
	current := s.first
	for current != nil {
		a += current.item.(string)
		a += " "
		current = current.next
	}
	return a
}

func main() {
	stack := new(Stack)
	scanner := bufio.NewScanner(os.Stdin)
	scanner.Split(bufio.ScanWords)
	for scanner.Scan() {
		item := scanner.Text()
		if item != "-" {
			stack.Push(item)
		} else if !stack.IsEmpty() {
			fmt.Printf("%s %s", stack.Pop(), "")
		}
	}
	fmt.Printf("%s%d %s", "(", stack.Len(), "left on stack)\n")
}
